File: Solutions/876_Middle_of_the_Linked_List/876_Middle_of_the_Linked_List.cpp
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        int cnt = 0;
        ListNode* mid = head;
        
        while(head != nullptr) {
            if(cnt & 1)
                mid = mid->next;
            
            ++cnt;
            head = head->next;
        }
        
        return mid;
    }
};

//Fastest
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;
        while(fast && fast->next){
            fast = fast->next->next;
            slow = slow->next;
        }
        return slow;
    }
};

File: Solutions/876_Middle_of_the_Linked_List/876_Middle_of_the_Linked_List.js
//Second fastest
var middleNode = function(head) {
    let list = []
    
    while(head!=null){
        list.push(head);
        head = head.next;
    }
    
    let res = Math.floor(list.length/2)
    return list[res]
}

//Fastest
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var middleNode = function(head) {
    var slow = head, fast = head;
    
    while (fast && fast.next) {
        fast = fast.next.next;
        slow = slow.next;
    }
    
    return slow;
};